feat(reporter) local & remote Endpoint improvements

kikito opened the pull request:
Fixes #55

In all the Spans generated by kong:

* The `localEndpoint` is `{ serviceName = "kong" }` in all spans.

The `remoteEndpoint` has changed as follows:

* On the request (root) span, it remains as before:
  * `remoteEndpoint.ipv4/v6` and `remoteEndpoint.port` point to
the consumer's forwarded ip and port

* On each balancer attempt span:
  * If a Kong Service was found when proxying, and that Service has a
`name`, then `remoteEndpoint.serviceName` will be the Service Name.
  * `remoteEndpoint.ipv4/v6` and `remoteEndpoint.port` will point to
the ip+port combination used on the span

* On the proxy span:
  * If a Kong Service was found when proxying, and that Service has a
`name`, then `remoteEndpoint.serviceName` will be the Service Name.
  * `remoteEndpoint.ipv4/v6` and `remoteEndpoint.port` will point to
the first successful combination of balancer attempts.

This PR also fixes a small issue where non-erroneous proxy spans appeared red on the UI because they the tag `error` set to `false`.

kikito commented:
I've got some concerns about this PR, quoting my comment in #55:

> * For all the spans created by the plugin, the localEndpoint is now `{ serviceName = "kong" }`, and nothing else (kong doesn't know its own IP address). It seems a bit ... unuseful?
> * When a span is tagged `error: false` it appears red on the UI. Should I remove the error tag and only use it when there's an error?
> * On the usual "API gateway" scenario, kong proxies traffic between a "downstream consumer" and an "upstream service", generating a root "request" span, with one proxy span and 1 or more balancer tries. These (now) all have the same `serviceName`, when Kong has identified a service to proxy to. But the ip4/port values can be very different: on the root span they point to the consumer, while on the other spans they point to the upstream service ips/ports. I think it is weird that they have the same serviceName but such different ips/ports.

bungle commented on kong/plugins/zipkin/opentracing.lua:
```diff
@@ -19,6 +19,22 @@ local OpenTracingHandler = {
+  elseif ctx.api and ctx.api.id then
+    span:set_tag("kong.api", ctx.api.id)
+  end
+end
```
Should we use `pdk`, or is that too big of a change considering that code `predates`(?) PDK?

e.g.
```lua
local service = kong.router.get_service()
local route = kong.router.get_route()
```

kikito replied:
Yes, done!

bungle commented:
Should we also update version:
https://github.com/Kong/kong-plugin-zipkin/blob/master/kong/plugins/zipkin/handler.lua#L9

jcchavezs commented:
> Otherwise remoteEndpoint = { serviceName = "<service.id>" } where id is a uuid.

I would rather not add anything if this is the case. The idea of having a `serviceName` is that we have something identifiable whereas a UUID can't be related to an actual service.

bungle commented:
> > Otherwise remoteEndpoint = { serviceName = "<service.id>" } where id is a uuid.
> 
> I would rather not add anything if this is the case. The idea of having a `serviceName` is that we have something identifiable whereas a UUID can't be related to an actual service.

Perhaps we could construct service `uri` and use that in case it does not have `name`.

jcchavezs commented:
That information is already in the tags. The remoteEndpoint.serviceName is
not too important as the downstream service will fill the gaps with its own
localEndpoint.serviceName and definitively not worth if it can't be
identifiable.

tor. 16. jan. 2020, 17:55 skrev Aapo Talvensaari <notifications@github.com>:

> Otherwise remoteEndpoint = { serviceName = "<service.id>" } where id is a
> uuid.
>
> I would rather not add anything if this is the case. The idea of having a
> serviceName is that we have something identifiable whereas a UUID can't
> be related to an actual service.
>
> Perhaps we could construct service uri and use that in case it does not
> have name.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/Kong/kong-plugin-zipkin/pull/63?email_source=notifications&email_token=AAXOYAUKHYCR27Y2WJGARY3Q6CGO3A5CNFSM4KGJNLLKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEJEYVCA#issuecomment-575244936>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAXOYAQMEXJSG2TKOGCWGCDQ6CGO3ANCNFSM4KGJNLLA>
> .
>

kikito commented:
I have removed the "serviceName can be the id" part. Only named services will appear on the remoteEndpoint. I have also replaced the use of ngx.ctx by kong-specific `kong.ctx.plugin` when possible (it was not possible in a lot of instances, because we needed things like `ngx.ctx.KONG_ACCESS_START` which aren't there).

jcchavezs commented:
LGTM, any input @adriancole?

adriancole commented on spec/zipkin_spec.lua:
```diff
@@ -168,22 +173,35 @@ describe("integration tests with zipkin server [#" .. strategy .. "]", function(
       ["http.status_code"] = "200", -- found (matches server status)
+      ["kong.route"] = route.id,
+      ["kong.service"] = route.service.id,
+      ["kong.consumer.port"] = consumer_port,
```
if this is a server span, and the consumer is the client, then this ip/port can move to remoteEndpoint vs being tags.

I'm not sure the concept of consumer or if this is a server span, so if this isn't the case, ignore

kikito replied:
> if this is a server span, and the consumer is the client, then this ip/port can move to remoteEndpoint vs being tags.

I think that is what we had before. Let me see if I have understood correctly.

On an API gateway scenario (the most common one for Kong at the moment) Kong acts as a proxy between a consumer (downstream) and a Service (upstream).

```
Request:
Consumer   --->   Kong  --->  Service

Response:
Consumer   <---   Kong  <---  Service
```

The *root span* generated by the zipkin plugin (kind = server) starts when Kong receives the consumer request, and ends when kong sends the response back to the consumer. So this span's remoteEndpoint's entity should point to the consumer.

Kong generates more spans while it communicates with the Service (at least 2, kind = client) and the Consumer isn't involved there. You are saying that the remoteEndpoint entity on those should be the Service.

On the other hand, `localEndpoint` should always be Kong.

Did I understand this correctly?

adriancole replied:
I agree that only the direct remote span should add the remoteEndpoint. In other words, it doesn't propagate to span data in children and so forth. I think what was confusing me was presence of the consumer tags.

I guess you are copying these to other spans for some reason. At least in zipkin this isn't required for search because search is trace scoped. Copying tags to children (about the consumer/client in this case) would add index load and increase the data size. If were me, if adding additional tags about the consumer/client I'd put in a comment as to why to prevent questions like this later.

kikito replied:
Hi, I apologize, but I have re-read your last comment and I still don't understand. Could you please spell out what you think `remoteEndpoint` should be on each of the spans? Right now all spans will "point" to the Service, and all `localEndpoints` point to Kong. Should I change anything (i.e. make root span's remoteEndpoint be the consumer) or not?

> I think what was confusing me was presence of the consumer tags.

The way I see it is that for Kong there isn't a single "remote entity", there are at least two (the consumer and the Service). I can put one or the other as a `remoteEntity`, but I thought it would be useful to add the other one somewhere. I don't know which one is more convenient or appropriate.

adriancole replied:
lets focus this on remoteEndpoint

a span of kind server, its remoteEndpoint is the caller which in RPC terms is the upstream client
similarly a span of kind server, its remoteEndpoint is who it is calling, in RPC terms the downstream server

as a span is either client or server and not both it should be unambiguous what to place into the remoteEndpoint without further need to clarify with tags. if there is an intermediate span between the server and the client on the same host in the same request it has no remoteendpoint and no need to carry any RPC related properties.

the additional terminology of consumer and service seem to be directionally oriented to me.. like labeling the two sides of the proxy and be the remoteEndpoint.serviceName of a client or server span depending on the direction. I dont understand why there would need to be other tags about this but I have sortof said all I think I can.

adriancole replied:
not sure it helps but in zipkin the endpoint.serviceName is the name of a node independent of direction. the labels of the node called or being called go in this field when known. that's where I think these consumer or service names should go, but I honestly dont know the subtleties which are making this confusing

jcchavezs replied:
I think what Adrian means is:

**For Consumer   --->   Kong  --->  Service**
- for `Consumer   --->   Kong` this `SERVER` span should have the `consumer IP` at `remoteEndpoint`
- for `Kong  --->  Service` this `CLIENT` span should have the `service IP` at `remoteEndpoint`

In either case `consumer IP` should not be a tag. Any other span that does not involve a remote call is a local span and that does not need a `remoteEndpoint`.

Are there any other spans that involve remote calls?

kikito replied:
> a span of kind server, its remoteEndpoint is the caller 

Ok, I will make remoteEndpoint point to the consumer on the root span, and to the kong service on the internal spans. And not use custom kong tags for this.

adriancole commented:
made one note. also I agree with @jcchavezs commentary from before. thanks for working on this, folks!

jcchavezs commented on kong/plugins/zipkin/opentracing.lua:
```diff
@@ -160,8 +174,8 @@ elseif subsystem == "stream" then
         ["span.kind"] = "server",
-        [ip_tag(forwarded_ip)] = forwarded_ip,
-        ["peer.port"] = kong.client.get_forwarded_port(),
+        ["kong.consumer." .. ip_tag(forwarded_ip)] = forwarded_ip,
```
I am not a lua expert. Is this going to be dynamic tag naming?

kikito replied:
I am going to remove this as per the conversation with @adriancole, but for the sake of completion: this would add a tag called either `"kong.consumer.ipv4"` or `"kong.consumer.ipv6"`, depending on the kind of IP.

jcchavezs replied:
I see. Good to know.

jcchavezs commented:
I see. Good to know.

tir. 21. jan. 2020, 16:53 skrev Enrique García Cota <
notifications@github.com>:

> *@kikito* commented on this pull request.
> ------------------------------
>
> In kong/plugins/zipkin/opentracing.lua
> <https://github.com/Kong/kong-plugin-zipkin/pull/63#discussion_r369084727>
> :
>
> > @@ -160,8 +174,8 @@ elseif subsystem == "stream" then
>        tags = {
>          component = "kong",
>          ["span.kind"] = "server",
> -        [ip_tag(forwarded_ip)] = forwarded_ip,
> -        ["peer.port"] = kong.client.get_forwarded_port(),
> +        ["kong.consumer." .. ip_tag(forwarded_ip)] = forwarded_ip,
>
> I am going to remove this as per the conversation with @adriancole
> <https://github.com/adriancole>, but for the sake of completion: this
> would add a tag called either "kong.consumer.ipv4" or "kong.consumer.ipv6",
> depending on the kind of IP.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Kong/kong-plugin-zipkin/pull/63?email_source=notifications&email_token=AAXOYAUQ643ER6WUOPKY6TLQ64LAJA5CNFSM4KGJNLLKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCSPVMDY#discussion_r369084727>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAXOYASLZH23YLFJFDPKKL3Q64LAJANCNFSM4KGJNLLA>
> .
>

kikito commented:
All requested changes have been done. I have updated the PR description accordingly.

adriancole commented:
:shipit: thanks for all the hard yards
